packages/shared/src/sql.ts

```diff
@@ -191,3 +192,90 @@ function stripCommentsAndStrings(sql: string): {
     parseError,
   };
 }
+
+type SqlResultChunkData = Pick<
+  SqlResultChunkInterface,
+  "columns" | "numRows" | "data"
+>;
+
+export function encodeSQLResults(
+  // Raw SQL results
+  results: Record<string, unknown>[],
+  // 4MB default chunk size (document max is 16MB, but leave plenty of room for overhead)
+  chunkSize: number = 4000000,
+): SqlResultChunkData[] {
+  if (results.length === 0) {
+    return [];
+  }
+
+  const columns = Object.keys(results[0]);
+  const encodedResults: SqlResultChunkData[] = [];
+
+  function createChunk(): SqlResultChunkData {
+    const chunk: SqlResultChunkData = {
+      columns,
+      numRows: 0,
+      data: {},
+    };
+    columns.forEach((col) => {
+      chunk.data[col] = [];
+    });
+    return chunk;
+  }
+
+  function getSize(value: unknown): number {
+    if (value === null || value === undefined) return 1;
+    if (typeof value === "boolean") return 1;
+    if (typeof value === "number") return 8;
+    if (typeof value === "string") return value.length + 5;
+
+    // Each nested field has overhead, so just multiply by 2 to be extra conservative
+    return 1 + JSON.stringify(value).length * 2;
```

Comment:
nit/issue: I don't think we have objects/arrays on many of the queries, but calling JSON.stringify for each row can potentially be super expensive.

I don't know a better solution at the moment though 🥲